### ISSUE TITLE
Replace stack / safestack

### ISSUE THREAD

**Issue by paulidale** [openssl/openssl#10657]
The current stack/safe_stack have a number of problems:
1. their API cannot be modified because they are public and
2. the data structure is overloaded by trying to be a vector, a list, a stack and a double ended queue (_deque_).

The solution is to have multiple data types each specialised for the purpose for which it is being used.  I.e. to have a vector, a linked list, a stack and a deque.  These data structures are different and have different performance trade-offs.  There is no one size fits all linear data structure.

After each data structure is implemented, the uses of safe_stack will need be analysed and converted as appropriate.

**Comment by richsalz** [comment]
There are *many* public API's that use a STACK that have to be considered as well.
I hope this is not being considered for 3.0

**Comment by paulidale** [comment]
Yes, in addition to stack and safestack being public.

**Comment by richsalz** [comment]
https://github.com/openssl/openssl/issues/10656#issuecomment-567316575

Especially if you're going to not revise the API.

**Comment by levitte** [comment]
Stack and safestack being public isn't a problem per se.

**Comment by paulidale** [comment]
The inline functions in the header file are a problem :)
The creeping (crept) API is a concern.

**Comment by levitte** [comment]
The direct *use* of the `DEFINE` macros in the public header file is the problem, yes.  The rest, not.

**Comment by paulidale** [comment]
I suspect it would be easier to create a new stack data structure than to deprecate and remove those `DEFINE` macros.  Kind of unfortunate, the stack is pretty good at being a stack.  It's the all other (mis-)uses it's been used for.  TLS cipher suite selection is the worst I remember seeing.

**Comment by richsalz** [comment]
> The direct _use_ of the `DEFINE` macros in the public header file is the problem, yes. The rest, not.

That's because we have to plays games with "static inline" to accomodate some platforms.

**Comment by levitte** [comment]
> That's because we have to plays games with "static inline" to accomodate some platforms.

Kinda sorta. If we had those DEFINE macro calls in the sources where they are used, we would have had zero problems.

**Comment by richsalz** [comment]
Except where the API has a stack.

**Comment by levitte** [comment]
No. STACK_OF(type) is safe from automatic inline function generation.

**Comment by richsalz** [comment]
Well, you have to replace ```DEFINE_STACK_OF`` with ``STACK_OF`` else you get compiler warnings.  The following diff shows how simple the changes are:
```
; g diff
diff --git a/crypto/asn1/asn_mime.c b/crypto/asn1/asn_mime.c
index 757fe9eb7a..71a06492eb 100644
--- a/crypto/asn1/asn_mime.c
+++ b/crypto/asn1/asn_mime.c
@@ -18,6 +18,8 @@
 #include "internal/bio.h"
 #include "asn1_local.h"
 
+DEFINE_STACK_OF(BIO)
+
 /*
  * Generalised MIME like utilities for streaming ASN1. Although many have a
  * PKCS7/CMS like flavour others are more general purpose.
diff --git a/crypto/conf/conf_def.c b/crypto/conf/conf_def.c
index 9718b73a18..9dbda10edf 100644
--- a/crypto/conf/conf_def.c
+++ b/crypto/conf/conf_def.c
@@ -27,6 +27,8 @@
 # endif
 #endif
 
+DEFINE_STACK_OF(BIO)
+
 #ifndef S_ISDIR
 # define S_ISDIR(a) (((a) & S_IFMT) == S_IFDIR)
 #endif
diff --git a/include/openssl/bio.h b/include/openssl/bio.h
index 50ac82cc31..d02ed9a406 100644
--- a/include/openssl/bio.h
+++ b/include/openssl/bio.h
@@ -285,7 +285,7 @@ int BIO_method_type(const BIO *b);
 typedef int BIO_info_cb(BIO *, int, int);
 typedef BIO_info_cb bio_info_cb;  /* backward compatibility */
 
-DEFINE_STACK_OF(BIO)
+STACK_OF(BIO);
 
 /* Prefix and suffix callback in ASN1 BIO */
 typedef int asn1_ps_func (BIO *b, unsigned char **pbuf, int *plen,
; 
```

I might do this as a code cleanup during holiday week.  Of course, if my doc PR's get merged then I can move onto FIPS work as promised....

**Comment by levitte** [comment]
Yeah, that's more or less what I'm saying.  However, that `DEFINE_STACK_BIO(BIO)` removal, while I technically agree with that 100%, it is a breaking change because some might actually expect to have `sk_BIO_*` just by including `<openssl/bio.h>`...  and that's the actual bullet that we haven't bitten yet.

**Comment by richsalz** [comment]
What will it take to come to a decision?  I think I can make it compatible if configured without no-deprecated, but it will be a a bit of a nuisance.

**Comment by t8m** [comment]
Breaking things only for no-deprecated would be acceptable I think.

**Comment by levitte** [comment]
We've had this discussion before. See #8102 and #8955...
FYI, deprecation makes zero difference.

**Comment by richsalz** [comment]
I came up with a work-around; PR coming in a few hours.

**Comment by richsalz** [comment]
See #10666

**Comment by rsbeckerca** [comment]
I'm not sure this helps one way or another - it probably doesn't - but the 3.0.0 definition of DEFINE_STACK_OF and related in safestack.h breaks under every C compiler on HPE NonStop platforms from c89 to c11 in the c-preprocessor. We get duplicate definitions of the inline methods. I have not worked out whether this is a bug in the preprocessor or the macros or some combination. Just when we thought we finally had things working (on 1.1.1g anyway). I can open a separate issue, but it seemed reasonable to mention it here. I might be able to help on a rework.

**Comment by paulidale** [comment]
A separate issue would be better, this is tracking something else.

**Comment by rsbeckerca** [comment]
Thanks @paulidale. I've opened it as #12667.

**Comment by beldmit** [comment]
Is this bug still relevant?

**Comment by paulidale** [comment]
I think it is still relevant.  It should not be done for 3.0.

**Comment by rsbeckerca** [comment]
I will keep testing :)

**Comment by t8m** [comment]
This is not going to be done any time soon and there is no point in having such kind of issues here for indefinite amount of time.

Any potential individual replacement of the stack with better structures needs to be handled separately. We are NOT going to replace all stack usage throughout the codebase at once, especially because it is also part of the library API calls in many cases.
